heat/core/io.py

```diff
@@ -354,36 +396,178 @@ def save_netcdf(data, path, variable, mode="w", **kwargs):
         is_split = data.split is not None
         _, _, slices = data.comm.chunk(data.gshape, data.split if is_split else 0)
 
+        def __get_expanded_split(shape, expandedShape, split):
+            """Returns the hypothetical split-axis of a dndarray of shape=shape and
+            split=split if it was expanded to expandedShape by adding empty dimensions.
+
+            Parameters
+            ----------
+            shape : tuple(int)
+                Shape of a dndarray.
+            expandedShape : tuple(int)
+                Shape of hypothetical expanded dndarray.
+            split : int or None
+                split-axis of dndarray.
+
+            Returns
+            -------
+            int
+                split-Axis of expanded dndarray.
+
+            Raises
+            -------
+            ValueError
+                If the shapes differ in non-empty dimensions.
+
+            """
+            # Get indices of non-empty dimensions and squeezed shapes
+            ind_nonempty, sq_shape = np.array([[i, v] for i, v in enumerate(shape) if v != 1]).T
+            ex_ind_nonempty, sq_ex = np.array(
+                [[i, v] for i, v in enumerate(expandedShape) if v != 1]
+            ).T
+            if not all(sq_shape == sq_ex):
+                raise ValueError(
+                    "Shapes %s and %s differ in non-empty dimensions" % (shape, expandedShape)
+                )
+            if len(shape) == len(expandedShape):  # actually not expanded at all
+                return split
+            if split is None:  # not split at all
+                return None
+            if split in ind_nonempty:  # split along non-empty dimension
+                split_sq = ind_nonempty.tolist().index(split)  # split-axis in squeezed shape
+                return ex_ind_nonempty[split_sq]
+            # split along empty dimension: split doesnt matter, only one process contains data
+            # return the last empty dimension (in expanded shape) before (the first nonempty dimension after split)
+            # number of nonempty elems before split
+            ne_before_split = split - shape[:split].count(1)
+            ind_ne_after_split = ind_nonempty[
+                ne_before_split
+            ]  # index of (first nonempty element after split) in squeezed shape
+            return max(
+                i
+                for i, v in enumerate(expandedShape[: max(ex_ind_nonempty[:ind_ne_after_split])])
+                if v == 1
+            )
+
+        def __merge_slices(var, var_slices, data, data_slices=None):
+            """ Using var[var_slices][data_slices] = data
+            combines a __getitem__ with a __setitem__ call, therefore it does not allow
+            parallelization of the write-operation and does not work is var_slices = slice(None)
```

Comment:
Not clear why parallelization would not work here. I am hesitant to merge this

Reply:
My reasoning is that because var_slices specifies the location of the (global) dndarray (and therefore is the same for every process), doing `var[var_slices]` loads the global data onto every single process as a ndarray, only to overwrite the local chunk using `[data_slices] = data`. I believe this does not allow parallelization because in the only method we call on the netcdf.Variable, the `__getitem__`, every process requests access to same location (var_slices).

Also, when I tried to use `var[var_slices][data_slices] = data`, the tests failed (in some cases, it did not write at all), so I came up with the idea to merge the var_slices and data_slices into a single set (tuple) of slices (or other indices).

Reply:
there was an update to the slicing recently. can you see if the new slicing solved your issue?

Reply:
Thanks, but I don't believe this is relevent here.

Maybe I didn't make it clear what this method actually does:
Each process knows its location inside the global dndarray via `comm.chunk()`. This method calculates (for each process) its location inside the (global) netCDF-File, such that every process can write to its corresponding location.  Therefore this method does not apply any slicing in HeAT.

Anyway, where can I find the changes made to slicing? I couldnt find them in the changelog

Reply:
Hi Ben,

`comm.chunk` also returns the offset of the local 0-index with respect to the global one, would  using `var.larray[var_slices-offset]`  allow you to parallelize the write operation? (if I understand the problem that is)

Reply:
Hi Claudia,
I don't understand why you would index the local array with the global offset? Also, please keep in mind that `var` is the `netcdf4.Variable`-object. I am now reworking the documentation for this method to clarify what it does

Reply:
Hopefully, this is more clear:

            This method allows replacing:
            ``var[var_slices][data_slices] = data``
            (a `netcdf4.Variable.__getitem__` and a `numpy.ndarray.__setitem__` call)
            with:
            ``var[ __merge_slices(var, var_slices, data, data_slices) ] = data``
            (a single `netcdf4.Variable.__setitem__` call)

            This is necessary because performing the former would, in the `__getitem__`, load the
            global dataset onto every process in local `numpy-ndarrays`. Then, the `__setitem__`
            would write the local `chunk` into the `numpy-ndarray`.

            The latter allows the netcdf4 library to parallelize the write-operation by directly
            using the `netcdf4.Variable.__setitem__` method.